flink-test-utils-parent/flink-test-utils-junit/src/test/java/org/apache/flink/testutils/junit/RetryOnExceptionExtensionTest.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.testutils.junit;
+
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.TestTemplate;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/** Tests for the RetryOnException annotation. */
+@ExtendWith(RetryExtension.class)
+public class RetryOnExceptionExtensionTest {
+
+    private static final int NUMBER_OF_RUNS = 3;
+
+    private static int runsForSuccessfulTest = 0;
+
+    private static int runsForTestWithMatchingException = 0;
+
+    private static int runsForTestWithSubclassException = 0;
+
+    private static int runsForPassAfterOneFailure = 0;
+
+    @AfterAll
+    public static void verify() {
+        assertEquals(NUMBER_OF_RUNS + 1, runsForTestWithMatchingException);
+        assertEquals(NUMBER_OF_RUNS + 1, runsForTestWithSubclassException);
+        assertEquals(1, runsForSuccessfulTest);
+        assertEquals(2, runsForPassAfterOneFailure);
+    }
+
+    @TestTemplate
```

Comment:
Is this always required if once wants to use `@RetryOnException?` Can we set it up somehow that the Retry annotations is sufficient?

Reply:
Yes, it is always required. If the annotation `RetryOnException` is annotated by `@TestTemplate`, the `@RetryOnException` is sufficient.
But both `RetryRule` and `RetryExtension` use `@RetryOnException` and `@RetryOnFailure`, it can not be easily changed.

Reply:
What would happen for juni4 tests if they were to be annotated as such?

Reply:
The test annotated with both `@TestTemplate` and `@Test` will execute twice in the junit jupiter engine and junit vintage engine. The test in junit jupiter will fail because of lacking the `TestTemplateInvocationContextProvider`.

Reply:
That's unfortunate :/ But that means once we migrated everything to junit4 we can still do it, correct?

Reply:
We still wouldn't wanna do that because we'd lose the ability to set up retries on the class level. (Because you cannot annotate classes with TestTemplate).

Reply:
Yes, `RetryOnException` annotated by `@TestTemplate` is not a good idea. It seems like that there could be both Junit4 test and Junit5 test in the same test class, and they will be executed in different engines.